File: rust/src/bin/permutations.rs
fn generate_permutations(n: usize, a: &mut [String]) {
    if n == 0 {
        println!("{}", a.join("\t"));
    } else {
        for i in 0..n {
            generate_permutations(n - 1, a);
            a.swap(if n % 2 == 0 { 0 } else { i }, n);
        }
        generate_permutations(n - 1, a);
    }
}

fn main() {
    let mut args: Vec<String> = std::env::args().skip(1).collect();
    generate_permutations(args.len() - 1, &mut args);
}

File: clojure/permutations.clj
(defn swap [a i j]
  (assoc a j (a i) i (a j)))

(defn print-permutations [v n]
  (if (zero? n)
    (do (println (clojure.string/join "\t" v)) v)
    (loop [i 0 a v]
      (if (< i n)
        (let [a (print-permutations a (dec n))]
          (recur (inc i) (swap a (if (even? n) 0 i) n)))
        (print-permutations a (dec n))))))

(print-permutations
  (vec *command-line-args*)
  (dec (count *command-line-args*)))
